src/pip/_internal/utils/unpacking.py

```diff
@@ -220,8 +218,10 @@ def untar_file(filename, location):
                     continue
                 ensure_dir(os.path.dirname(path))
                 with open(path, 'wb') as destfp:
-                    shutil.copyfileobj(fp, destfp)
-                fp.close()
+                    if fp:
+                        shutil.copyfileobj(fp, destfp)
+                if fp:
+                    fp.close()
```

Comment:
`assert fp` before should be sufficient to please mypy and is closer to what the current code does (i.e. fail loudly if fp is None).

Reply:
Added assert in https://github.com/pypa/pip/pull/8277/commits/50cbd6a03245d6315da74c470cf1c481ef554cba